src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/LLVMObjectWriter.cs

```diff
@@ -168,63 +174,122 @@ private unsafe void EmitObjectNode(ObjectNode node, ObjectData nodeContents)
             // Calculate the size of this object node.
             int dataSizeInBytes = nodeContents.Data.Length;
             int dataSizeInBytesAligned = dataSizeInBytes.AlignUp(pointerSize); // TODO-LLVM: do not pad out the data.
-            int dataSizeInPointers = dataSizeInBytesAligned / pointerSize;
+            int dataSizeInElements = dataSizeInBytesAligned / pointerSize;
 
-            // Create and initialize the LLVM global value.
-            LLVMTypeRef dataSymbolType = LLVMTypeRef.CreateArray(_ptrType, (uint)dataSizeInPointers);
-            LLVMValueRef dataSymbol = _module.AddGlobal(dataSymbolName, dataSymbolType);
-            dataSymbol.Section = node.GetSection(_nodeFactory).Name;
-            dataSymbol.Alignment = (uint)nodeContents.Alignment;
+            // If we need to create unaligned relocs then use a struct for the symbol.
+            // Start with the premise that most nodes will just contain aligned relocs.
+            bool useStruct = false;
+
+            if (_currentObjectData.Length < dataSizeInElements)
+            {
+                Array.Resize(ref _currentObjectData, Math.Max(_currentObjectData.Length * 2, dataSizeInElements));
+            }
+
+            Span<LLVMValueRef> dataElements = _currentObjectData.AsSpan(0, dataSizeInElements);
+            ReadOnlySpan<byte> data = nodeContents.Data.AsSpan();
+
+            // Indicies in byte units to allow us to "zip" the binary data with the relocs
+            int dataOffset = 0;
+            int relocIndex = 0;
+            int elementOffset = 0;
 
-            // Emit the value of this object node.
-            if (_currentObjectData.Length < dataSizeInPointers)
+            int relocLength = nodeContents.Relocs.Length;
+            int nextRelocOffset = 0;
+            bool nextRelocValid = relocIndex < relocLength;
+            if (nextRelocValid)
             {
-                Array.Resize(ref _currentObjectData, Math.Max(_currentObjectData.Length * 2, dataSizeInPointers));
+                nextRelocOffset = nodeContents.Relocs[0].Offset;
             }
-            Span<LLVMValueRef> dataElements = _currentObjectData.AsSpan(0, dataSizeInPointers);
-            dataElements.Clear();
 
-            // Emit relocations. We assume these are always aligned.
-            foreach (Relocation reloc in nodeContents.Relocs)
+            while (dataOffset < dataSizeInBytes)
             {
-                long delta;
-                fixed (void* location = &nodeContents.Data[reloc.Offset])
+                if (!useStruct && nextRelocValid && nextRelocOffset % pointerSize != 0)
                 {
-                    delta = Relocation.ReadValue(reloc.RelocType, location);
+                    // Switch from array to struct.
```

Comment:
```suggestion
                    // Switch from array to struct. This will need more elements because binary data is output byte-by-byte.
```
Took me a little to realize why the size recalculation is needed.